### ISSUE TITLE
MPI based h5py not compatible with python3.10 

### ISSUE THREAD

**Issue by jorgensd** [h5py/h5py#2100]
The version of mpi4py required by h5py [https://github.com/h5py/h5py/blob/master/setup.py#L48](3.0.3) [does not](https://github.com/mpi4py/mpi4py/blob/3.0.3/setup.py#L77) have Python 3.10 support.
Later versions of mpi4py supports Python 3.10, see: https://pypi.org/project/mpi4py/

Ubuntu 22.04 LTS is shipped with python 3.10 as default.

 * Ubuntu 22.04 LTS
 * Python version 3.10
 * Ubuntu package manager
 *  h5py version (e.g. 3.6.0)
 * HDF5 version (e.g. 1.10.7)
 * Minimal Docker environment:
```bash
FROM ubuntu:22.04

WORKDIR /tmp

ARG MPI="mpich"

RUN export DEBIAN_FRONTEND=noninteractive && \
    apt-get -qq update && \
    apt-get -yq --with-new-pkgs -o Dpkg::Options::="--force-confold" upgrade && \
    apt-get -y install \
	python3-dev \
    libhdf5-${MPI}-dev \
    lib${MPI}-dev \
    pkg-config \
    python3-pip && \
    apt-get -y install \
    git  && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*


RUN HDF5_MPI="ON" HDF5_DIR="/usr/lib/x86_64-linux-gnu/hdf5/mpich/" CC=mpicc pip3 install --no-cache-dir --no-binary=h5py h5py 
```
* Trace :
Gives a very long error message, so I've only added the beginning, where it is clear that the issue is due to the restriction 

```bash
 Collecting h5py
   Downloading h5py-3.6.0.tar.gz (384 kB)
      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 384.2/384.2 kB 70.6 MB/s eta 0:00:00
   Installing build dependencies: started
   Installing build dependencies: finished with status 'done'
   Getting requirements to build wheel: started
   Getting requirements to build wheel: finished with status 'done'
   Installing backend dependencies: started
   Installing backend dependencies: still running...
   Installing backend dependencies: still running...
   Installing backend dependencies: finished with status 'error'
   error: subprocess-exited-with-error

   × pip subprocess to install backend dependencies did not run successfully.
 │ exit code: 1
  ╰─> [7884 lines of output]
       WARNING: The directory '/github/home/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.
       Collecting mpi4py==3.0.3
            ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 19.4 MB/s eta 0:00:00
         Preparing metadata (setup.py): started
         Preparing metadata (setup.py): finished with status 'done'
       Building wheels for collected packages: mpi4py
         Building wheel for mpi4py (setup.py): started
         Building wheel for mpi4py (setup.py): still running...
         Building wheel for mpi4py (setup.py): finished with status 'error'
         error: subprocess-exited-with-error
```

**Comment by tacaswell** [comment]
It looks like we are pinning back to the oldest version of mpi4py to forward ABI compatibly.  I think the fix is to add a py310 condition to that list with the minimum mpi4py that works.  

Could you do that @jorgensd ?

**Comment by rbeucher** [comment]
I came across the issue today. That's a pb. Not sure which is the minimum `mpi4py` though

```SETUP_REQUIRES.append("mpi4py ==3.1.3; python_version>='3.10'")```

**Comment by tacaswell** [comment]
I can get mpi4py 3.1.0 to install on py310 from pypi, but I do not have any mpi workloads to actually test on.

The frustrating thing about this is that I think mpi4py 3.0.3 will work just fine on py310, however because mpi4py includes a cython output file in their sdist on pypi it is not install able.  If you pull from the tag in github it will install:

```
✔ 22:40:56 $ python -m pip install git+https://github.com/mpi4py/mpi4pi@3.0.3
Collecting git+https://github.com/mpi4py/mpi4pi@3.0.3
  Cloning https://github.com/mpi4py/mpi4pi (to revision 3.0.3) to /tmp/pip-req-build-y0ma4k2i
  Running command git clone --filter=blob:none --quiet https://github.com/mpi4py/mpi4pi /tmp/pip-req-build-y0ma4k2i
Username for 'https://github.com': ^CERROR: Operation cancelled by user
WARNING: You are using pip version 22.0.4; however, version 22.1 is available.
You should consider upgrading via the '/tmp/clean/bin/python -m pip install --upgrade pip' command.
(clean) ~ |✓
✘ 22:41:54 $ python -m pip install git+https://github.com/mpi4py/mpi4py@3.0.3
Collecting git+https://github.com/mpi4py/mpi4py@3.0.3
  Cloning https://github.com/mpi4py/mpi4py (to revision 3.0.3) to /tmp/pip-req-build-usn_1wlc
  Running command git clone --filter=blob:none --quiet https://github.com/mpi4py/mpi4py /tmp/pip-req-build-usn_1wlc
  Running command git checkout -q 086de047f90652249f0a73d77cdad19b873a3ab4
  Resolved https://github.com/mpi4py/mpi4py to commit 086de047f90652249f0a73d77cdad19b873a3ab4
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: mpi4py
  Building wheel for mpi4py (setup.py) ... done
  Created wheel for mpi4py: filename=mpi4py-3.0.3-cp310-cp310-linux_x86_64.whl size=3011353 sha256=dbde16603fa6ae4c6fce328b43191ca919371a187126ca83167c63b45d924b40
  Stored in directory: /tmp/pip-ephem-wheel-cache-rdp_w09n/wheels/24/81/81/7652c4171221d7c996d72d3d9f434a4fd024a50f2df4b1958a
Successfully built mpi4py
Installing collected packages: mpi4py
Successfully installed mpi4py-3.0.3
WARNING: You are using pip version 22.0.4; however, version 22.1 is available.
You should consider upgrading via the '/tmp/clean/bin/python -m pip install --upgrade pip' command.
(clean) ~ |✓
✔ 22:42:35 $ python --version
Python 3.10.4
(clean) ~ |✓


```

**Comment by rbeucher** [comment]
I think mpi4py is fine. I have been using 3.1.3 for a while with python 3.10.

**Comment by takluyver** [comment]
Closed by #2101.
